src/main/java/org/jenkinsci/plugins/workflow/steps/EnvironmentExpander.java

```diff
@@ -41,13 +45,32 @@
  * Pass into {@link BodyInvoker#withContext}.
  */
 public abstract class EnvironmentExpander implements Serializable {
+    private Set<String> watchedVars;
 
     /**
      * May add environment variables to a context.
      * @param env an original set of environment variables
      */
     public abstract void expand(@Nonnull EnvVars env) throws IOException, InterruptedException;
 
+    public void watch(String var) {
+        if (watchedVars == null) {
+            watchedVars = new HashSet<>();
+        }
+        watchedVars.add(var);
+    }
+    public void watchAll(Collection<?  extends String> vars) {
+        if (watchedVars == null) {
+            watchedVars = new HashSet<>();
+        }
+        watchedVars.addAll(vars);
+    }
+
+    @CheckForNull
+    public Set<String> getWatchedVars() {
+        return watchedVars;
+    }
```

Comment:
I would prefer to add only this method and have it return `Collections.emptySet()` by default, and have implementations override it as needed. `credentials-binding` and other implementors of `EnvironmentExpander` can add their own fields/methods for storing state as needed, for example `BindingStep.Overrider.getWatchedVars` might be able to just `return overrides.keySet()`, and `MergedEnvironmentExpander` can implement `getWatchedVars` as something like:

```
return Streams.of(original.getWatchedVars(), subsequent.getWatchedVars())
        .flatMap(Set::stream)
        .collect(Collectors.toSet())
```

As far as naming, my preference would be to use "secrets" in the name since "watched" is pretty generic, so I would call it something like `getSecretVars`, but I would see what other reviewers think.

Reply:
That makes sense to me. Naming I'm fine with `secrets`. I only did `watched` to try and stay as generic as possible in case this could be used as some other way.

Reply:
Also as far as naming, I just remembered that this API is analogous to [`AbstractBuild.getSensitiveBuildVariables`](https://github.com/jenkinsci/jenkins/blob/87f94049984a5b6a802ae07d86edb0825c869457/core/src/main/java/hudson/model/AbstractBuild.java#L935) (for build types that extend `AbstractBuild`, e.g. almost everything but Pipeline), so maybe good to use "sensitive" instead of "secret" for consistency.

Regarding trying to keep this generic, if you wanted to do that, I think you would need to add a parameter to distinguish the use case, maybe something like this: `Set<String> getWatchedVars(String watchType)`, and then we'd call `getWatchedVars(WatchType.SECRET)` or something. Since we don't have any other use cases right now, I would probably not make this generic, but if needed in the future we could add a new method and have `getWatchedVars()` delegate to it.